proposals/cel_generating_policies.md

```diff
@@ -670,3 +670,90 @@ There are three main approaches to solve this:
 
       3. Maintain a cache of registered GVRs to prevent duplication.
 
+### Lazy Evaluation on First Execution
+
+Workflow:
+
+1. A user creates a GeneratingPolicy.
+
+2. A trigger resource is created that matches the GeneratingPolicy.
+
+3. An UpdateRequest (UR) is created to generate the downstream resource.
+
+4. The background controller processes the UR and evaluates the CEL expressions.
+
+5. During the first execution of `generator.Apply(...)`, we call `watchManager.SyncWatchers()`, passing the policy name and the list of the generated resources.
+
+    ```go
+    type Resource struct {
+      Name      string
+      Namespace string
+      Hash      string
+      Labels    map[string]string
+      Data      *unstructured.Unstructured
+    }
+
+    type watcher struct {
+      watcher       watch.Interface
+      metadataCache map[types.UID]Resource
+    }
+
+    type WatchManager struct {
+      // clients
+      client dclient.Interface
+
+      // mapper
+      restMapper meta.RESTMapper
+
+      // dynamicWatchers refrences the GVR of the generated resources to the corresponding watcher.
+      dynamicWatchers map[schema.GroupVersionResource]*watcher
+      // policyRefs maps the policy name to the set of generated resources.
+      policyRefs map[string][]schema.GroupVersionResource
+      // refCount tracks the number of policies that generates the same resource.
+      refCount map[schema.GroupVersionResource]int
+
+      log  logr.Logger
+      lock sync.Mutex
+    }
+    ```
+
+    - The `SyncWatchers()` does the following:
+    
+        a. For each new GVR, it starts a dynamic watcher. To be efficient, it uses a reference counter (refCount). If two policies both generate ConfigMaps, only one watcher for ConfigMaps is created.
+
+        b. It compares the new set of GVRs with the old set for that policy. If a policy is updated to no longer generate a certain resource kind, the reference counter for that GVR is decremented. If the count reaches zero, the corresponding watcher is stopped.
+
+        c. It populates an internal `metadataCache` with details (name, namespace, hash) of the downstream resources.
+
+    - Once a watcher is running for a resource kind (e.g., ConfigMap), it receives events for all resources of that kind cluster-wide.
+
+    - On UPDATE event:
+
+        a. The `handleUpdate` function checks if the updated resource is one of the downstream resources it manages (by looking it up in its `metadataCache`).
+        
+        b. If it is a downstream resource, it calculates the hash of the incoming resource and compares it to the hash stored in the cache.
+        
+        c. If the hashes differ, it means a user has modified the resource. The WatchManager then reverts the change by re-applying the original, desired state from its cache. This is done by getting the original resource from the cache.
```

Comment:
Do we need to store the origianl resource in the cache? Would it impact performance in large scales?

Reply:
Yup, we need to store the original resource in the cache because we revert the downstream changes by retrieving the original data from the cache.